curriculum/challenges/english/25-front-end-development/lecture-working-with-data-fetching-and-memoization-in-react/67d2f51ff2c927713caa24fa.md

```diff
@@ -2,13 +2,185 @@
+  };
+
+  useEffect(() => {
+    console.log("useEffect runs");
+  }, [handleClick]);
+
+  return (
+    <div>
+      <p>Count: {count}</p>
+      <button onClick={handleClick}>Increment</button>
+    </div>
+  );
+}
+
+export default Counter;
+```
+
+In the component, the effect runs any time `handleClick` changes because the `handleClick` function is being recreated on every render.
+
+To fix this, you need to tell React to treat the `handleClick` function as the same thing across renders by memoizing it with the `useCallback` hook, so it doesn't get recreated:
+
+```js
+import { useState, useEffect, useCallback } from "react";
+
+function Counter() {
+  const [count, setCount] = useState(0);
+
+  // Memoize the handleClick function with useCallback
+  const handleClick = useCallback(() => {
+    setCount((prevCount) => prevCount + 1);
+  }, []);
+
+  useEffect(() => {
+    console.log("useEffect runs");
+  }, [handleClick]);
+
+  return (
+    <div>
+      <p>Count: {count}</p>
+      <button onClick={handleClick}>Increment</button>
+    </div>
+  );
+}
+
+export default Counter;
+```
+
+Now the `handleClick` function is not being recreated on every render:
+
+To show you how the `React.memo` (or `memo`) higher-order function works and the `useCallback` hook work in tandem, here's a `Counter` component with a `handleClick` function that needs `useCallback` but is currently not using it:
+
+```js
+import { useState, useEffect, useCallback } from "react";
+import CounterChild from "./CounterChild";
+
+function Counter() {
+  const [count, setCount] = useState(0);
+  const [timer, setTimer] = useState(new Date().toLocaleTimeString());
+
+  const handleClick = () => {
+    setCount(count + 1);
+  };
+
+  useEffect(() => {
+    const interval = setInterval(() => {
+      setTimer(new Date().toLocaleTimeString());
+    }, 1000);
+
+    return () => clearInterval(interval);
+  }, []);
+
+  return (
+    <div>
+      <h1>Time: {timer}</h1>
+      <p>Count: {count}</p>
+      <button onClick={handleClick}>Increment</button>
+      <CounterChild onClick={handleClick} />
+    </div>
+  );
+}
+
+export default Counter;
+```
+
+This function also has a timer in state that updates every second. This makes the component re-render every time the `timer` changes, making the `handleClick` function get recreated on every render.
+
+That's why the `handleClick` needs to be memoized with `useCallback`.
+
+Here's the `CounterChild` component:
+
+```js
+const CounterChild = ({ onClick }) => {
+  console.log("CounterChild component rendered");
+  return <button onClick={onClick}>Increment from Child</button>;
+};
+
+export default CounterChild;
+```
+
+This `CounterChild` component takes an `onClick` prop, giving you the ability to also increment the counter from it.
+
+Since the `CounterChild` is a child of the `Counter` component, it will also render any time the `Counter` re-renders due to the changing timer. So, the `CounterChild` also needs to be memoized.
```

Comment:
```suggestion
Since the `CounterChild` component is a child of the `Counter` component, it will also render any time the `Counter` re-renders due to the changing timer. So, the `CounterChild` also needs to be memoized.
```